File: Roadmap/03 - ESTRUCTURAS DE DATOS/javascript/AndresMCardenas.js
/* 
  * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
 *

  """

# Creación de estructuras de datos

# Listas
*/

let lista = [1, 2, 3, 4, 5, 6, 7, 8, 9];

// Inserción
lista.push(10);
console.log(lista);

// Borrado

lista.pop();
console.log(lista);

// Actualización

lista[0] = 0;
console.log(lista);

// Ordenación

lista.sort();
console.log(lista);

// Diccionarios

let diccionario = {
  nombre: "Andres",
  apellido: "Cardenas",
  edad: 30,
};

// Inserción

diccionario["sexo"] = "Masculino";
console.log(diccionario);

// Borrado

delete diccionario["sexo"];
console.log(diccionario);

// Actualización

diccionario["edad"] = 31;
console.log(diccionario);

// Ordenación
 
// No aplica

// Tuplas

let tupla = [1, "Andres", 30];
console.log(tupla);

// Inserción

tupla.push("Cardenas");
console.log(tupla);

// Borrado

tupla.pop();
console.log(tupla);

// Actualización

tupla[0] = 0;
console.log(tupla);

// Ordenación

tupla.sort();
console.log(tupla);

// Conjuntos son estructuras de datos que no permiten elementos repetidos

let conjunto = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9]);
console.log(conjunto);

// Inserción

conjunto.add(10);
console.log(conjunto);

// Borrado

conjunto.delete(10);
console.log(conjunto);

// Actualización no aplica dado que no permite dadp que no se tiene accesos a los elementos por indice

// Ordenación no aplica dado que no permite se tiene accesos a los elementos por indice

// Diccionarios ordenados

let diccionarioOrdenado = new Map([
  ["nombre", "Andres"],
  ["apellido", "Cardenas"],
  ["edad", 30],
]);

console.log(diccionarioOrdenado);

// busqueda

console.log(diccionarioOrdenado.get("nombre"));


// Inserción

diccionarioOrdenado.set("sexo", "Masculino");
console.log(diccionarioOrdenado);

// Borrado

diccionarioOrdenado.delete("sexo");
console.log(diccionarioOrdenado);

// Actualización

diccionarioOrdenado.set("edad", 31);
console.log(diccionarioOrdenado);

// Ordenación no se aplica dado que no tiene indice

// Pilas

let pila = [1, 2, 3, 4, 5, 6, 7, 8, 9];
console.log(pila);

// Inserción

pila.push(10);
console.log(pila);

// Borrado

pila.pop();
console.log(pila);

// Actualización

pila[0] = 0;
console.log(pila);

// Ordenación

pila.sort();
console.log(pila);

// Colas

let cola = [1, 2, 3, 4, 5, 6, 7, 8, 9];
console.log(cola);

// Inserción

cola.push(10);
console.log(cola);

// Borrado

cola.shift();
console.log(cola);

// Actualización

cola[0] = 0;
console.log(cola);

// Ordenación

cola.sort();
console.log(cola);

// Árboles son estructuras de datos no lineales que se componen de nodos o vértices y ramas u hojas

class Nodo {
  constructor(valor) {
    this.valor = valor;
    this.hijos = [];
  }
}

class Arbol {
  constructor(valor) {
    this.raiz = new Nodo(valor);
  }
}

let arbol = new Arbol(1);
console.log(arbol);

// Inserción

arbol.raiz.hijos.push(new Nodo(2));
console.log(arbol);

// Borrado

arbol.raiz.hijos.pop();
console.log(arbol);

// Actualización

arbol.raiz.valor = 0;
console.log(arbol);

// Ordenación no aplica dado que no tiene indice

// Grafos son estructuras de datos no lineales que se componen de nodos o vértices y aristas

// class Nodo {
//   constructor(valor) {
//     this.valor = valor;
//     this.hijos = [];
//   }
// }

// class Grafo {
//   constructor(valor) {
//     this.raiz = new Nodo(valor);
//   }
// }

// let grafo = new Grafo(1);
// console.log(grafo);

// Inserción

// grafo.raiz.hijos.push(new Nodo(2));
// console.log(grafo);

// // Borrado

// grafo.raiz.hijos.pop();
// console.log(grafo);

// // Actualización

// grafo.raiz.valor = 0;
// console.log(grafo);

// Ordenación no aplica dado que no tiene indice

// Matrices son estructuras de datos que se componen de filas y columnas

let matriz = [
  [1, 2, 3],
  [4, 5, 6],
];
console.log(matriz);

// Inserción

matriz.push([7, 8, 9]);
console.log(matriz);

// Borrado

matriz.pop();
console.log(matriz);

// Actualización

matriz[0][0] = 0;
console.log(matriz);

// Ordenación

matriz.sort();
console.log(matriz);

/*
"""
  * DIFICULTAD EXTRA(opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización
  * y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar,
 * y a continuación los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no númericos y con más
  * de 11 dígitos(o el número de dígitos que quieras).
 * - También se debe proponer una operación de finalización del programa.

"""
*/

File: Roadmap/03 - ESTRUCTURAS DE DATOS/python/AndresMCardenas.py
"""
 * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
 *
 
"""

# Creación de estructuras de datos

# Listas

listaNumeros = [1, 2, 3, 4, 5] # Lista de números
print(listaNumeros)

listaString = ["a", "b", "c", "d", "e"] # Lista de strings
print(listaString)

listaMixta = [1, "a", 2, "b", 3, "c"] # Lista mixta
print(listaMixta)

# Tuplas

tuplaNumeros = (1, 2, 3, 4, 5) # Tupla de números
print(tuplaNumeros)

tuplaString = ("a", "b", "c", "d", "e") # Tupla de strings
print(tuplaString)

tuplaMixta = (1, "a", 2, "b", 3, "c") # Tupla mixta
print(tuplaMixta)

# Diccionarios

diccionarioNumeros = {1: "uno", 2: "dos", 3: "tres", 4: "cuatro", 5: "cinco"} # Diccionario de números
print(diccionarioNumeros)

diccionarioString = {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5} # Diccionario de strings
print(diccionarioString)

diccionarioMixto = {1: "a", "b": 2, 3: "c", "d": 4, "e": 5} # Diccionario mixto
print(diccionarioMixto)

#buscar en diccionario
print(diccionarioNumeros[1])

#insertar en diccionario
diccionarioNumeros[6] = "seis"
print(diccionarioNumeros)

#actualizar en diccionario
diccionarioNumeros[6] = "SEIS"
print(diccionarioNumeros)

#eliminar en diccionario
del diccionarioNumeros[6]
print(diccionarioNumeros)


# Operaciones de inserción, borrado, actualización y ordenación

# Listas

listaNumeros.append(6) # Insertar un elemento al final de la lista
print(listaNumeros)

listaNumeros.insert(0, 0) # Insertar un elemento en una posición específica de la lista
print(listaNumeros)

listaNumeros.pop() # Eliminar el último elemento de la lista
print(listaNumeros)

listaNumeros.pop(0) # Eliminar un elemento en una posición específica de la lista
print(listaNumeros)

listaNumeros[0] = 1 # Actualizar un elemento en una posición específica de la lista
print(listaNumeros)

listaNumeros.sort() # Ordenar la lista
print(listaNumeros)

# Tuplas

# Las tuplas son inmutables, por lo que no se pueden modificar

# Diccionarios

diccionarioNumeros[6] = "seis" # Insertar un elemento en el diccionario
print(diccionarioNumeros)

diccionarioNumeros[6] = "SEIS" # Actualizar un elemento en el diccionario
print(diccionarioNumeros)

del diccionarioNumeros[6] # Eliminar un elemento en el diccionario
print(diccionarioNumeros)

"""
* DIFICULTAD EXTRA (opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización
 *   y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar,
 *   y a continuación los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no númericos y con más
 *   de 11 dígitos (o el número de dígitos que quieras).
 * - También se debe proponer una operación de finalización del programa.

"""

# Agenda de contactos

agenda = {} # Diccionario vacío

control = True

opcion = 0

# Funciones de la agenda de contactos 
      
# Buscar contacto      
def buscarContacto():
  nombre = input("Introduzca el nombre del contacto: ").upper()
  if nombre in agenda:
    print("El número de teléfono de", nombre, "es", agenda[nombre])
  else:
    print("El contacto", nombre, "no existe")

# Insertar contacto
def insertarContacto():
  nombre = input("Introduzca el nombre del contacto: ").upper()
  if nombre in agenda:
    print("El contacto", nombre, "ya existe")
  else:
    numero = input("Introduzca el número de teléfono del contacto: ")
    if numero.isnumeric() and len(numero) > 11:
      agenda[nombre] = numero
      print("El contacto", nombre, "ha sido añadido")
    else:
      print("El número de teléfono introducido no es válido")

# Actualizar contacto
def actualizarContacto():
  nombre = input("Introduzca el nombre del contacto: ").upper()
  if nombre in agenda:
    numero = input("Introduzca el nuevo número de teléfono del contacto con el indicador del pais: ")
    if numero.isnumeric() and len(numero) > 11:
      agenda[nombre] = numero
      print("El contacto", nombre, "ha sido actualizado")
    else:
      print("El número de teléfono introducido no es válido")
  else:
    print("El contacto", nombre, "no existe")

# Eliminar contacto
def eliminarContacto():
  nombre = input("Introduzca el nombre del contacto: ").upper()
  if nombre in agenda:
    del agenda[nombre]
    print("El contacto", nombre, "ha sido eliminado")
  else:
    print("El contacto", nombre, "no existe")

# Mostrar agenda
def mostrarAgenda():
  print("Agenda de contactos")
  for nombre, numero in agenda.items():
    print(nombre, ":", numero)

# Salir
def salir():
  global control
  control = False
  print("Fin del programa")

# Ejecución de la agenda de contactos
while control:
  print("----------------------------------------")
  print("Agenda de contactos")
  print("----------------------------------------")
  print("1. Buscar contacto")
  print("2. Insertar contacto")
  print("3. Actualizar contacto")
  print("4. Eliminar contacto")
  print("5. Mostrar agenda")
  print("6. Salir")
  print("----------------------------------------")
  opcion = int(input("¿Qué operación desea realizar? "))
  print("----------------------------------------")
  
  if opcion == 1:
    buscarContacto()
  elif opcion == 2:
    insertarContacto()
  elif opcion == 3:
    actualizarContacto()
  elif opcion == 4:
    eliminarContacto()
  elif opcion == 5:
    mostrarAgenda()
  elif opcion == 6:
    salir()
  else:
    print("Opción incorrecta")



  






















